src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,902 @@
+ * properties:
+ *  - it will have only one shard,
+ *  - its distribution method will be DISTRIBUTE_BY_NONE,
+ *  - its replication model will be ReplicationModel,
+ *  - its replication factor will be set to 1.
+ * Similar to reference tables, it has only 1 placement. In addition to that, that
+ * single placement is only allowed to be on the coordinator.
+ */
+static void
+CreateCitusLocalTable(Oid relationId)
+{
+	/* these checks should be done before acquiring the lock on the table */
+	EnsureCoordinator();
+	EnsureTableOwner(relationId);
+
+	/*
+	 * Lock target relation with an AccessExclusiveLock as we don't want
+	 * multiple backends manipulating this relation. We could actually simply
+	 * lock the relation without opening it. However, we also want to check
+	 * if the relation does not exist or dropped by another backend. Also,
+	 * we open the relation with try_relation_open instead of relation_open
+	 * to give a nice error in case the table is dropped by another backend.
+	 */
+	Relation relation = try_relation_open(relationId, AccessExclusiveLock);
+
+	ErrorIfUnsupportedCreateCitusLocalTable(relation);
+
+	/*
+	 * We immediately close relation with NoLock right after opening it. This is
+	 * because, in this function, we may execute ALTER TABLE commands modifying
+	 * relation's column definitions and postgres does not allow us to do so when
+	 * the table is still open. (See the postgres function CheckTableNotInUse for
+	 * more information)
+	 */
+	relation_close(relation, NoLock);
+
+
+	ObjectAddress tableAddress = { 0 };
+	ObjectAddressSet(tableAddress, RelationRelationId, relationId);
+
+	/*
+	 * Ensure dependencies first as we will create shell table on the other nodes
+	 * in the MX case.
+	 */
+	EnsureDependenciesExistOnAllNodes(&tableAddress);
+
+	/*
+	 * Make sure that existing reference tables have been replicated to all
+	 * the nodes such that we can create foreign keys and joins work
+	 * immediately after creation.
+	 */
+	EnsureReferenceTablesExistOnAllNodes();
+
+	List *shellTableDDLEvents = GetShellTableDDLEventsForCitusLocalTable(relationId);
+
+	char *relationName = get_rel_name(relationId);
+	Oid relationSchemaId = get_rel_namespace(relationId);
+
+	/* below we convert relation with relationId to the shard relation */
+	uint64 shardId = ConvertLocalTableToShard(relationId);
+
+	/*
+	 * As we retrieved the DDL commands necessary to create the shell table
+	 * from scratch, below we simply recreate the shell table executing them
+	 * via process utility.
+	 */
+	CreateShellTableForCitusLocalTable(shellTableDDLEvents);
+
+	/*
+	 * Set shellRelationId as the relation with relationId now points
+	 * to the shard relation.
+	 */
+	Oid shellRelationId = get_relname_relid(relationName, relationSchemaId);
+
+	/* assert that we created the shell table properly in the same schema */
+	Assert(OidIsValid(shellRelationId));
+
+	Oid shardRelationId = relationId;
```

Comment:
As we already have a similar comment for `Oid shellRelationId`, moving `Oid shardRelationId` just above `Oid shellRelationId` might be good to prevent comment duplication.